### ISSUE TITLE
UBSan check in __attribute__((const))

### ISSUE THREAD

**Issue by Teemperor** [llvm/llvm-project#55225]
The following reduced example shifts by 2000 (which is clearly UB), but when this happens inside a `__attribute__((const))` function then an optimized program doesn't report this (or fail) with -fsanitize=undefined.

```c++
__attribute__((const))
int getShift() {
  return 2000;
}

__attribute__((const))
int hasUB() {
  return 1 << getShift();
}

int main() {
    return hasUB();
}
```

Example here:
https://godbolt.org/z/dKccK99dd

Note that the `__attribute__((const))` is actually suggested by GCC here for the function containing UB.

```c++
source>: In function 'int hasUB()':
<source>:7:5: warning: function might be candidate for attribute 'const' [-Wsuggest-attribute=const]
    7 | int hasUB() {
      |     ^~~~~
```

**Comment by vfdff** [comment]
it maybe not an issue, as gcc will also not report info related to  -fsanitize=undefined (gcc optimized that), see detail in  https://godbolt.org/z/vc4GvEs9o

**Comment by Teemperor** [comment]
Maybe godbolt is glitching out for me, but isn't that the same situation (Clang removes the check, GCC keeps it)

**Comment by fhahn** [comment]
cc @AaronBallman @yln @rjmccall 

Any thoughts about whether the attributes Clang generates are in line with the guarantees by `__attribute__(const)`? Or any suggestions on who could know?

**Comment by AaronBallman** [comment]
Oofda, that's a fun one. My understanding of `const` and `pure` is that both of them have to be idempotent (along with other constraints) so that each evaluation is guaranteed to result in the same returned value. Once your function has UB in it, there's no guarantee that it's idempotent any longer because who knows what it will do on any given invocation. So I see that example as having UB twice over, after a fashion: the first is the shift amount and the second is marking a function `const` when it doesn't meet the requirements. So I think LLVM is free to do whatever it wants to that function in that case. (Seems like `pure` has the same behavior in LLVM: https://godbolt.org/z/ao414fWjs)

That said, it sure seems like the behavior we *want* is for UBSan to catch (both) of these issues.

**Comment by Teemperor** [comment]
FWIW, I fixed that in my downstream LLVM but just ignoring these attributes in the frontend (so, if any sanitizer is enabled the frontend doesn't emit the function attributes in the IR). I believe there is no use case we can break by ignoring these attributes (aside from the performance impact which seems negligible given the overall sanitizer overhead).

**Comment by AaronBallman** [comment]
> I believe there is no use case we can break by ignoring these attributes (aside from the performance impact which seems negligible given the overall sanitizer overhead).

I also don't know of any use cases broken by ignoring those attributes, aside from performance.

**Comment by rjmccall** [comment]
This is definitely okay to have done with `__attribute__((const))` and `__attribute__((pure))`, which are basically defined functionally: they mean that it's okay to arbitrarily hoist, combine, and remove calls to them, subject only to the restriction (for `pure`) that they can depend on state.

I don't understand how LLVM is removing the call to `hasUB`, though.  It's `readnone willreturn`, but its result is used.

**Comment by fhahn** [comment]
> I don't understand how LLVM is removing the call to hasUB, though. It's readnone willreturn, but its result is used. Is LLVM also extracting the return value out of the call despite the UB checks within the function? Maybe that's an interesting path to explore.


I think the key for removing the call is that Clang also adds the `nounwind` attribute to `hasUB`. So all paths must exit through a return: https://llvm.godbolt.org/z/jTbhYTr73

**Comment by rjmccall** [comment]
No, but the return value is used.  You can't just remove calls to a `readnone` function whose return value is used; it's not a "never call me at all, this isn't real code" attribute, you have to actually produce the right result.

**Comment by fhahn** [comment]
> No, but the return value is used. You can't just remove calls to a readnone function whose return value is used; it's not a "never call me at all, this isn't real code" attribute, you have to actually produce the right result. The return value from hasUB is observed as the status result of the program. So LLVM must also be extracting the return value from the callee independent of the later optimization to remove the call.

I might be missing something, but the return value is known (`poison`). Passes like `-ipsccp` will replace all uses with the known return value. If the attributes allow, the call will then be removed (as is the case here)

**Comment by rjmccall** [comment]
Yeah, I'm just saying that that's also implicated here: we're pulling information out of a `readnone` call in a way that allows the call to be removed completely, in this case eliminating a UB check.

**Comment by nikic** [comment]
@rjmccall I'm still not sure what you're trying to say here in actionable terms. The transform is perfectly fine *given the IR LLVM has received*. If Clang instrumentation makes the function non-readnone (as sanitizer error reporting obviously does), then it falls to Clang to omit placing the attribute.

**Comment by rjmccall** [comment]
That is probably the easiest fix for this problem, and in practice it will usually be fine.  It does mean that e.g. assertions in `const` functions are effectively not guaranteed to run in any way even if the result of the call is used, and that seems like a very limiting thing to say to users.

What I am saying is this: Clang has to have some way to translate the `const` and `pure` attributes, and outside of intrinsics these are the primary things that turn into the LLVM `readnone` and `readonly` attributes, at least in C.  If you read the GCC documentation, those attributes are clearly defined operationally in terms of the optimizations they enable: users are prohibited from doing things within `const` and `pure` functions to the exact extent that those things would break under CSE-like optimizations.  In particular, `const` and `pure` functions are allowed to have internal effects that do not break under CSE, such as memoization and assertions, and the most natural interpretation of the operational definition of these attributes is that those internal effects are required to occur (somewhere) on any path that depends on the call result.  Under that interpretation, pulling the return value out of the call and then deleting the call completely, when you haven't proven the absence of those allowed internal effects, would be a miscompile.

**Comment by fhahn** [comment]
> Under that interpretation, pulling the return value out of the call and then deleting the call completely, when you haven't proven the absence of those allowed internal effects, would be a miscompile.

So from this interpretation it would be valid for the function to unwind and/or not returning, right? If that's the case, Clang probably shouldn't add `nouwind`/`will return` to those functions in general. This will prevent removing the call, unless LLVM can prove the function will return and not unwind.

**Comment by davidstone** [comment]
If I'm understanding correctly, a suggestion was to ignore these attributes if the program is being compiled under sanitizers? One downside of this approach is the interaction with sanitizers + `[[assume(condition)]]`. I expect sanitizers to report an incorrect assumption (UBSan claims to catch undefined behavior, and violating an assumption is undefined behavior). However, if evaluating `condition` evaluates a function that is not at least `[[gnu::pure]]`, the assumption is stripped out entirely. If we ignore the attributes when compiled by sanitizers, we make it impossible to sanitize certain sources of undefined behavior.
